tamr_client/operation.py

```diff
@@ -0,0 +1,195 @@
+            results that would be produced by the operation are already up-to-date."""
+        resource_json = {
+            "id": "-1",
+            "type": "NOOP",
+            "description": _description,
+            "status": {
+                "state": "SUCCEEDED",
+                "startTime": _never,
+                "endTime": _never,
+                "message": "",
+            },
+            "created": {"username": "", "time": _never, "version": "-1"},
+            "lastModified": {"username": "", "time": _never, "version": "-1"},
+            "relativeId": "operations/-1",
+        }
+    else:
+        resource_json = response.json()
+    _id = resource_json["id"]
+    _url = URL(instance=instance, path=f"operations/{_id}")
+    return _from_json(_url, resource_json)
+
+
+def apply_options(
+    session: Session, operation: Operation, *, asynchronous: bool = False, **options
+) -> Operation:
+    """Applies operation options to this operation.
+
+    **NOTE**: This function **should not** be called directly. Rather, options should be
+    passed in through a higher-level function e.g. :func:`~tamr_client.dataset.unified.commit`.
+
+    Synchronous mode:
+        Automatically waits for operation to resolve before returning the
+        operation.
+
+    asynchronous mode:
+        Immediately return the ``'PENDING'`` operation. It is
+        up to the user to coordinate this operation with their code via
+        :func:`~tamr_client.operation.wait` and/or
+        :func:`~tamr_client.operation.poll` .
+
+    Args:
+        asynchronous: Whether or not to run in asynchronous mode. Default: ``False``.
+        ``**options``: When running in synchronous mode, these options are
+                passed to the underlying :func:`~tamr_client.operation.wait` call.
+    """
+    if asynchronous:
+        return operation
+    return wait(session, operation, **options)
+
+
+def _from_url(session: Session, url: URL) -> Operation:
+    """Get operation by URL
+
+    Fetches operation from Tamr server
+
+    Args:
+        url: Operation URL
+
+    Raises:
+        OperationNotFound: If no operation could be found at the specified URL.
+            Corresponds to a 404 HTTP error.
+        requests.HTTPError: If any other HTTP error is encountered.
+    """
+    r = session.get(str(url))
+    if r.status_code == 404:
+        raise NotFound(str(url))
+    data = response.successful(r).json()
+    return _from_json(url, data)
+
+
+def _from_json(url: URL, data: JsonDict):
+    """Make operation from JSON data (deserialize)
+
+    Args:
+        url: Operation URL
+        data: Operation JSON data from Tamr server
+    """
+    cp = deepcopy(data)
+    return Operation(
+        url,
+        id=cp["id"],
+        type=cp["type"],
+        status=cp.get("status"),
+        description=cp.get("description"),
+    )
+
+
+def poll(session: Session, operation: Operation) -> Operation:
```

Comment:
nit: move the intended interface functions to the top, and the `_`-prefixed functions to the bottom of this file